src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -130,20 +133,38 @@ def _make_candidate_from_link(
         name,  # type: Optional[str]
         version,  # type: Optional[_BaseVersion]
     ):
-        # type: (...) -> Candidate
+        # type: (...) -> Optional[Candidate]
         # TODO: Check already installed candidate, and use it if the link and
         # editable flag match.
+        if link in self._build_failures:
+            return None
         if template.editable:
             if link not in self._editable_candidate_cache:
-                self._editable_candidate_cache[link] = EditableCandidate(
-                    link, template, factory=self, name=name, version=version,
-                )
+                try:
+                    self._editable_candidate_cache[link] = EditableCandidate(
+                        link, template, factory=self,
+                        name=name, version=version,
+                    )
+                except InstallationError as e:
+                    if isinstance(e, HashError):
+                        raise
```

Comment:
We have to exclude HashError here so it can correctly emit the “wrong hash” error message, otherwise we’d skip the candidate with incorrect hash and emit “no matching candidate” instead. This is pretty bad, but unfortuantely generic build errors return plain InstallationError and we don’t have another way to do this without modifying the preparer.